typing/shape.mli

```diff
@@ -55,14 +55,19 @@
     when trying to jump to an identifier's declaration or definition. They are
     stored to that effect in the [uid_to_decl] table of cmt files. *)
 module Uid : sig
+  type intf_or_impl = Intf | Impl
   type t = private
     | Compilation_unit of string
-    | Item of { comp_unit: string; id: int }
+    | Item of { comp_unit: string; id: int; from: intf_or_impl }
     | Internal
     | Predef of string
 
   val reinit : unit -> unit
 
+  (** [set_from] allow changing the [from] information stored with uids
+    indicating if they emanate from an interface or an implementation file. *)
+  val set_from : intf_or_impl -> unit
```

Comment:
I think it would be better to store this information in the same place we store the current unit name (i.e. in `Env`). Possibly it would be worth creating a `current_unit` type containing the name and the `intf_or_impl` and then have `mk` take one of those and then have `Env.get_current_unit` return one of those and replace the vast majority of `Env.get_unit_name`s with `Env.get_current_unit`. (Also maybe rename `get_unit_name` to `get_current_unit_name` for clarity).

Reply:
It seems like a reasonable thing to do, but I am going to face circular dependencies issues if the `current_unit` type is defined in `Env` and refered to in `Shape.Uid`. I dug a bit more and there might be an even more appropriate place for that type definition: the `Unit_info` module introduced by @Octachron. Currently, the unit_name in env is set from the corresponding field of `Unit_info.t`:
```ocaml
let with_info ~native ~tool_name ~source_file ~output_prefix ~dump_ext k =
  Compmisc.init_path ();
  let target = Unit_info.make ~source_file output_prefix in
  Env.set_unit_name (Unit_info.modname target);
  let env = Compmisc.initial_env() in
  let dump_file = String.concat "." [output_prefix; dump_ext] in
  Compmisc.with_ppf_dump ~file_prefix:dump_file @@ fun ppf_dump ->
  k {
    target;
    env;
    ppf_dump;
    tool_name;
    native;
  }
  ```

Maybe we could simply add the `impl_or_intf` field to the `Unit_info.t` and store all this in the `Env` reference: `Env.set_unit_info target` ?

Reply:
I made changes in that direction in [0bb4209](https://github.com/ocaml/ocaml/pull/13286/commits/0bb42093634a73f936aecf137dc4cde262f1f98a)